src/app/data-client.ts

```diff
@@ -709,6 +714,157 @@ export class DataClient {
+
+    const metadata = new dataSyncPb.UploadMetadata();
+    metadata.setPartId(partId);
+    metadata.setComponentType(componentType);
+    metadata.setComponentName(componentName);
+    metadata.setMethodName(methodName);
+    metadata.setType(dataSyncPb.DataType.DATA_TYPE_TABULAR_SENSOR);
+    metadata.setTagsList(tags ?? []);
+
+    const sensorContents: dataSyncPb.SensorData[] = [];
+    for (const [i, data] of tabularData.entries()) {
+      const sensorData = new dataSyncPb.SensorData();
+
+      const sensorMetadata = new dataSyncPb.SensorMetadata();
+      const dates = dataRequestTimes[i];
+      if (dates) {
+        if (dates[0]) {
+          sensorMetadata.setTimeRequested(Timestamp.fromDate(dates[0]));
+        }
+        if (dates[1]) {
+          sensorMetadata.setTimeReceived(Timestamp.fromDate(dates[1]));
+        }
+      }
+      sensorData.setMetadata(sensorMetadata);
+      sensorData.setStruct(googleStructPb.Struct.fromJavaScript(data));
+
+      sensorContents.push(sensorData);
+    }
+
+    const req = new dataSyncPb.DataCaptureUploadRequest();
+    req.setMetadata(metadata);
+    req.setSensorContentsList(sensorContents);
+
+    const response = await promisify<
+      dataSyncPb.DataCaptureUploadRequest,
+      dataSyncPb.DataCaptureUploadResponse
+    >(service.dataCaptureUpload.bind(service), req);
+    return response.getFileId();
+  }
+
+  /**
+   * Uploads the content and metadata for binary data.
+   *
+   * Upload binary data collected on a robot through a specific component (e.g.,
+   * a motor) along with the relevant metadata to app.viam.com. binary data can
+   * be found under the "Sensors" subtab of the Data tab on app.viam.com.
+   *
+   * @param binaryData The data to be uploaded, represented in bytes
+   * @param partId The part ID of the component used to capture the data
+   * @param componentType The type of the component used to capture the data
+   *   (e.g., "movementSensor")
+   * @param componentName The name of the component used to capture the data
+   * @param methodName The name of the method used to capture the data.
+   * @param fileExtension The file extension of binary data including the
+   *   period, e.g. .jpg, .png, .pcd. The backend will route the binary to its
+   *   corresponding mime type based on this extension. Files with a .jpeg,
+   *   .jpg, or .png extension will be saved to the images tab.
+   * @param tags The list of tags to allow for tag-based filtering when
+   *   retrieving data
+   * @param dataRequestTimes Containing `Date` objects denoting the times this
+   *   data was requested[0] by the robot and received[1] from the appropriate
+   *   sensor.
+   * @returns The file ID of the uploaded data
+   */
+  async binaryDataCaptureUpload(
+    binaryData: Uint8Array,
+    partId: string,
+    componentType: string,
+    componentName: string,
+    methodName: string,
+    fileExtension: string,
+    tags?: string[],
+    dataRequestTimes?: Date[]
```

Comment:
Similar comments as above re: making this a tuple and whether or not we need to check below for the existence of particular entries.